### ISSUE TITLE
Moving to "FermiOcean3'

### ISSUE THREAD

**Issue by stepan-tsirkin** [wannier-berri/wannier-berri#67]
Dear @manxkim , @jaemolihm 
may I bring your attention to the following [branch](https://github.com/wannier-berri/wannier-berri/tree/formula+berry-boltzmann)

Earlier I anounced the "FermiOcean" implementation of evaluation of fermisea quantities, now we are working on its 3rd version, and this one looks good. 

The Idea is to implement our building blocks as [nonabelian formulas](https://github.com/wannier-berri/wannier-berri/blob/formula%2Bberry-boltzmann/wannierberri/__formulas_nonabelian_3.py) and then we may construct their [FormulaProduct](https://github.com/wannier-berri/wannier-berri/blob/formula%2Bberry-boltzmann/wannierberri/__fermiocean3.py)s and integrate them weighted by the distribution function(or its derivatives) optionally involving tetrahedron method. The same Formulas may also be used for tabulation. 
In tyurn, the `Data_K` should be cleaned up, removing many of the routines. Basically it should only evaluate the "Hbar" quantities (i.e. doing FFTs and unitary rotations", not explicitly evaluating the Berry curvature and other "high-level" stuff.

So, this is going to be the major machinary for calculation of transport properties.

 When you have time, could you look at the implementation, and think if the implementations of kubo.py can be brought to this paradigm. As a start, could you implement the evaluation of spin-Berry curvature and spin-velocity matrix elements, at least for one fixed frequency (or at least for Omega=0) . This will, in particular open way to spin Nernst effect, and also will allow plotting (see #35) . Further we may think how to efficiently generalize the code for multiple frequencies.

Of course, we will not remove working code until we have a better working code, so this transition may take time, and there might be two implementations  side-by-side for some time.

**Comment by jaemolihm** [comment]
It will be great to integrate spin-related quantities to this new framework. I've only skimmed through the code yet, I will look deeper soon.

So, to implement the spin Berry curvature and spin Hall conductivity, will the first step be to add e.g. `class SpinOmega` in `__formulas_nonabelian_3.py` and `def spin_hall` in `__fermiocean3.py`?

**Comment by stepan-tsirkin** [comment]
Kind of . Starting from eq 1 of Ryoo paper
![Screenshot_2021-07-16-17-28-06-580_com google android apps docs](https://user-images.githubusercontent.com/22172977/125971996-5fae45ab-036a-412e-92ba-712f0fc43f3c.jpg)

you first need to implement a formula for the matrix element of j^a_b and for the matrix element if v.  Then, you may implement SpinOmega (for beginning, maybe just at zero frequency), and then spin_hall in __fermi_ocean_3.py.


And you may have several implementations for that matrix element, and the choice should be a parameter of spin_hall

One comment needed for the methods `nn` and `ln` . That notation comes from a manuscript that we are writing about gauge covariant matrices and their derivatives. 

 We divide the space into 'inner' and 'outer' states, and the matrix should be covariant under unitary transformatiins inside each if the states. 
Here `n` stands for an inner space index, and 'l' for the outer. nl and nn are evaluated differently in general.

**Comment by jaemolihm** [comment]
I finally started looking the code closely. One quick comment is that `super(CurrentClass, self)` is python2 syntax and in python3 one can just write `super()`.
https://stackoverflow.com/questions/21768391/how-can-i-call-super-so-its-compatible-in-2-and-3

**Comment by jaemolihm** [comment]
Another question is that is `Omega.nn` correct only for m=n or also for m/=n?

**Comment by jaemolihm** [comment]
@stepan-tsirkin I implemented static SHC to FermiOcean3. See https://github.com/jaemolihm/wannier-berri/tree/formula_shc.
I also added a test that kubo routine (with `omega=0` and a very small `eta=1e-10`) and FermiOcean3 gives the same SHC: see `tests/test_shc.py`.

How should I proceed? Should I make a PR from my branch to formula+berry-boltzmann?

By the way, the test was broken for that branch because of "Quantity ... is not available for integrate" errors.
